xbmc/filesystem/BlurayDirectory.cpp

```diff
@@ -66,49 +66,540 @@ std::string CBlurayDirectory::GetDiscInfoString(DiscInfo info)
 {
   switch (info)
   {
-  case XFILE::CBlurayDirectory::DiscInfo::TITLE:
-  {
-    if (!m_blurayInitialized)
-      return "";
-    const BLURAY_DISC_INFO* disc_info = bd_get_disc_info(m_bd);
-    if (!disc_info || !disc_info->bluray_detected)
-      return "";
+    case XFILE::CBlurayDirectory::DiscInfo::TITLE:
+    {
+      if (!m_blurayInitialized)
+        return "";
+      const BLURAY_DISC_INFO* disc_info = bd_get_disc_info(m_bd);
+      if (!disc_info || !disc_info->bluray_detected)
+        return "";
 
-    std::string title = "";
+      std::string title = "";
 
-#if (BLURAY_VERSION > BLURAY_VERSION_CODE(1,0,0))
-    title = disc_info->disc_name ? disc_info->disc_name : "";
+#if (BLURAY_VERSION > BLURAY_VERSION_CODE(1, 0, 0))
+      title = disc_info->disc_name ? disc_info->disc_name : "";
 #endif
 
-    return title;
+      return title;
+    }
+    case XFILE::CBlurayDirectory::DiscInfo::ID:
+    {
+      if (!m_blurayInitialized)
+        return "";
+
+      const BLURAY_DISC_INFO* disc_info = bd_get_disc_info(m_bd);
+      if (!disc_info || !disc_info->bluray_detected)
+        return "";
+
+      std::string id = "";
+
+#if (BLURAY_VERSION > BLURAY_VERSION_CODE(1, 0, 0))
+      id = disc_info->udf_volume_id ? disc_info->udf_volume_id : "";
+
+      if (id.empty())
+      {
+        id = HexToString(disc_info->disc_id, 20);
+      }
+#endif
+
+      return id;
+    }
+    default:
+      break;
   }
-  case XFILE::CBlurayDirectory::DiscInfo::ID:
+
+  return "";
+}
+
+static constexpr unsigned int MIN_EPISODE_LENGTH = 15 * 60; // 15 minutes
+static constexpr unsigned int MAX_EPISODE_DIFFERENCE = 30; // 30 seconds
+
+void CBlurayDirectory::GetEpisodeTitles(const CFileItem& episode,
+                                        CFileItemList& items,
+                                        const std::vector<CVideoInfoTag>& episodesOnDisc)
+{
+  // Get all titles on disc
+  // Sort by playlist for grouping later
+  CFileItemList allTitles;
+  GetTitles(false, allTitles);
+  SortDescription sorting;
+  sorting.sortBy = SortByFile;
+  allTitles.Sort(sorting);
+
+  // Find our episode on disc
+  unsigned int episodeOffset = 0;
+  for (int i = 0; i < (int)episodesOnDisc.size(); ++i)
   {
-    if (!m_blurayInitialized)
-      return "";
+    if (episodesOnDisc[i].m_iEpisode == episode.GetVideoInfoTag()->m_iEpisode)
+    {
+      episodeOffset = i;
+      break;
+    }
+  }
+
+  const unsigned int numEpisodes = (int)episodesOnDisc.size();
```

Comment:
Signed /unsigned mismatch and...
Please, no c-stype casts! `(int)`-> `static_cast<unsigned int>(...)`